rfc/rfc-51/rfc-51.md

```diff
@@ -148,20 +152,46 @@ hudi_cdc_table/
 
 Under a partition directory, the `.log` file with `CDCBlock` above will keep the changing data we have to materialize.
 
-There is an option to control what data is written to `CDCBlock`, that is `hoodie.table.cdc.supplemental.logging`. See the description of this config above.
+#### Persisting CDC in MOR: Write-on-indexing vs Write-on-compaction
+
+2 design choices on when to persist CDC in MOR tables:
+
+Write-on-indexing allows CDC info to be persisted at the earliest, however, in case of Flink writer or Bucket
+indexing, `op` (I/U/D) data is not available at indexing.
+
+Write-on-compaction can always persist CDC info and achieve standardization of implementation logic across engines,
+however, some delays are added to the CDC query results. Based on the business requirements, Log Compaction (RFC-48) or
+scheduling more frequent compaction can be used to minimize the latency.
 
-Spark DataSource example:
+The semantics we propose to establish are: when base files are written, the corresponding CDC data is also persisted.
+
+- For Spark
+  - inserts are written to base files: the CDC data `op=I` will be persisted
+  - updates/deletes that written to log files are compacted into base files: the CDC data `op=U|D` will be persisted
+- For Flink
+  - inserts/updates/deletes that written to log files are compacted into base files: the CDC data `op=I|U|D` will be
+    persisted
+
```

Comment:
>inserts/updates/deletes that written to log files are compacted into base files: the CDC data `op=I|U|D` will be
    persisted
    
I don't think the compaction generated cdc logs makes any sense in production, it lost the data freshness for cdc stream and it relies on the compaction service which it self is not a very robust infrastructure.

Reply:
What do you suggest on when we write out CDC log blocks when we use bucket index? 
Also why is compaction service not robust?

Reply:
> What do you suggest on when we write out CDC log blocks when we use bucket index

Does the index type (bloom filter/bucket) affect the CDC log blocks here ? Personally i think generating the log blocks during write on the fly is not very feasible because looking up the history records index would cost much which reduce the write throughput.

We can generate the cdc on the fly for the mor table reader, and the CDC log blocks generated by compaction is a optimization view like the cow table.

Reply:
@danny0405 looks like we're aligned here: the proposal here is to make use of compaction generate CDC log blocks. For flink MOR write, no index lookup will be or can be performed anyway. For spark MOR write, index lookup happens anyway regardless of enabling CDC (this depends on index type).

Reply:
That's true, but by default flink does not take the look up(for better throughput), and we may generate the changelog for reader on the fly.

Reply:
discussed offline we're aligned on this approach that flink write won't produce cdc log block